SSA/Projects/DCE/DCE.lean

```diff
@@ -0,0 +1,492 @@
+import SSA.Core.Framework
+
+/-- Delete a variable from a list. -/
+def Ctxt.delete (Γ : Ctxt Ty) (v : Γ.Var α) : Ctxt Ty :=
+  Γ.removeNth v.val
+
+/-- Witness that Γ' is Γ without v -/
+def Deleted {α : Ty} (Γ: Ctxt Ty) (v : Γ.Var α) (Γ' : Ctxt Ty) : Prop :=
+  Γ' = Γ.delete v
+
+/-- build a `Deleted` for a `(Γ.snoc α) → Γ`-/
+def Deleted.deleteSnoc (Γ : Ctxt Ty) (α : Ty) : Deleted (Γ.snoc α) (Ctxt.Var.last Γ α) Γ := rfl
+
+theorem List.removeNth_zero : List.removeNth (List.cons x xs) 0 = xs := rfl
+
+theorem List.removeNth_succ : List.removeNth (List.cons x xs) (.succ n) = x :: List.removeNth xs n := rfl
+
+/- removing from `xs ++ [x]` at index `(length xs)` equals `xs`. -/
+theorem List.removeNth_eq_len_snoc (hn : n = xs.length) : List.removeNth (xs ++ [x]) n = xs := by
```

Comment:
Also, why not `List.removeNth (xs ++ [x]) xs.length` I have a feeling it's because there's some `motive is not type correct issue` somewhere. But that's why you should try to use morphisms of contexts and not equalities.

Reply:
The style with `n` is easier to rewrite with, as it creates another proof obligation to show that `<something> = xs.length` instead of failing on the rewrite. I prefer to have terms that are "linear" (ie, no repeated patterns) and proof obligations to show that stuff that needs to be equal is equal.

Reply:
I changed the signature as requested.